agent/command/s3_copy.go

```diff
@@ -188,43 +238,76 @@ func (c *s3copy) s3Copy(ctx context.Context,
-			S3SourcePath:        s3CopyFile.Source.Path,
-			S3DestinationRegion: s3CopyFile.Destination.Region,
-			S3DestinationBucket: s3CopyFile.Destination.Bucket,
-			S3DestinationPath:   s3CopyFile.Destination.Path,
-			S3DisplayName:       s3CopyFile.DisplayName,
-			S3Permissions:       s3CopyFile.Permissions,
-		}
-
-		responseString, err := comm.S3Copy(ctx, td, &s3CopyReq)
+		s3CopyReq.AwsKey = c.AwsKey
+		s3CopyReq.AwsSecret = c.AwsSecret
 
-		if responseString != "" {
-			logger.Task().Infof("s3Copy response: %s", responseString)
+		// Now copy the file into the permanent location
+		srcOpts := pail.S3Options{
+			Credentials: pail.CreateAWSCredentials(s3CopyReq.AwsKey, s3CopyReq.AwsSecret, ""),
+			Region:      s3CopyReq.S3SourceRegion,
+			Name:        s3CopyReq.S3SourceBucket,
+			Permissions: pail.S3Permissions(s3CopyReq.S3Permissions),
 		}
 
+		srcBucket, err := pail.NewS3MultiPartBucketWithHTTPClient(client, srcOpts)
 		if err != nil {
-			err = errors.Wrap(err, "s3 push copy failed")
-			logger.Execution().Error(err)
-
-			if s3CopyFile.Optional {
-				logger.Execution().Errorf("file '%s' is optional, continuing",
-					s3CopyFile.DisplayName)
-				continue
-			} else {
-				return errors.Wrapf(err, "failed to push %s to %s",
-					s3CopyFile.Source.Path, s3CopyFile.Destination.Bucket)
-			}
-
+			connectionErr := errors.Wrap(err, "S3 copy failed, could not establish connection to source bucket")
+			logger.Task().Error(connectionErr)
+			return connectionErr
 		}
 
-		err = c.attachFiles(ctx, comm, logger, td, s3CopyReq)
+		if err := srcBucket.Check(ctx); err != nil {
+			return errors.Wrap(err, "invalid bucket")
+		}
+		destOpts := pail.S3Options{
+			Credentials: pail.CreateAWSCredentials(s3CopyReq.AwsKey, s3CopyReq.AwsSecret, ""),
+			Region:      s3CopyReq.S3DestinationRegion,
+			Name:        s3CopyReq.S3DestinationBucket,
+			Permissions: pail.S3Permissions(s3CopyReq.S3Permissions),
+		}
+		destBucket, err := pail.NewS3MultiPartBucket(destOpts)
 		if err != nil {
-			return errors.WithStack(err)
+			connectionErr := errors.Wrap(err, "S3 copy failed, could not establish connection to destination bucket")
+			logger.Task().Error(connectionErr)
+			return connectionErr
+		}
+	retryLoop:
+		for i := 0; i < maxS3OpAttempts; i++ {
+			select {
+			case <-ctx.Done():
+				return errors.New("s3 copy operation canceled")
+			case <-timer.C:
+				copyOpts := pail.CopyOptions{
+					SourceKey:         s3CopyReq.S3SourcePath,
+					DestinationKey:    s3CopyReq.S3DestinationPath,
+					DestinationBucket: destBucket,
+				}
+				err = srcBucket.Copy(ctx, copyOpts)
+				if err != nil {
+					newPushLog.Status = pushLogFailed
+					if err := comm.UpdatePushStatus(ctx, td, newPushLog); err != nil {
+						return errors.Wrap(err, "updating pushlog status failed for task")
+					}
+					if s3CopyFile.Optional {
+						logger.Execution().Errorf("S3 push copy failed to copy '%s' to '%s'. File is optional, continuing \n error: %v",
```

Comment:
Should this verbosely print the error with `%v` or just the error string `%s`? Seems like a user wouldn't want to see things like the stack trace you woudl get if you use `%v`.

Reply:
thanks, fixed